pmd-doc/src/main/java/net/sourceforge/pmd/docs/DeadLinksChecker.java

```diff
@@ -0,0 +1,327 @@
+            }
+        }
+        return joined;
+    }
+
+
+    private void addDeadLink(Map<Path, List<Future<String>>> fileToDeadLinks, Path file, Future<String> line) {
+        fileToDeadLinks.computeIfAbsent(file, k -> new ArrayList<>()).add(line);
+    }
+
+
+    private Set<String> extractLinkTargets(List<Path> mdFiles) {
+        final Set<String> htmlPages = new HashSet<>();
+        for (Path mdFile : mdFiles) {
+            final String pageContent = fileToString(mdFile);
+
+            // extract the permalink header field
+            final Matcher permalinkMatcher = MD_HEADER_PERMALINK.matcher(pageContent);
+            if (!permalinkMatcher.find()) {
+                continue;
+            }
+
+            final String pageUrl = permalinkMatcher.group(1)
+                                                   .replaceAll("^/+", ""); // remove the leading "/"
+
+            // add the root page
+            htmlPages.add(pageUrl);
+
+            // add all captions as anchors
+            final Matcher captionMatcher = MD_CAPTION.matcher(pageContent);
+            while (captionMatcher.find()) {
+                final String anchor = captionMatcher.group(1)
+                                                    .toLowerCase(Locale.ROOT)
+                                                    .replaceAll("[^a-z0-9_]+", "-") // replace all non-alphanumeric characters with dashes
+                                                    .replaceAll("^-+|-+$", ""); // trim leading or trailing dashes
+
+                htmlPages.add(pageUrl + "#" + anchor);
+            }
+        }
+        return htmlPages;
+    }
+
+
+    private List<Path> listMdFiles(Path pagesDirectory) {
+        final List<Path> mdFiles = new ArrayList<>();
+        try {
+            Files.walk(pagesDirectory)
+                 .filter(Files::isRegularFile)
+                 .filter(path -> path.toString().endsWith(".md"))
+                 .forEach(mdFiles::add);
+        } catch (IOException ex) {
+            throw new RuntimeException("error listing files in " + pagesDirectory, ex);
+        }
+        return mdFiles;
+    }
+
+
+    private String fileToString(Path mdFile) {
+        try (InputStream inputStream = Files.newInputStream(mdFile)) {
+            return IOUtils.toString(inputStream, Charset.forName("UTF-8"));
+        } catch (IOException ex) {
+            throw new RuntimeException("error reading " + mdFile, ex);
+        }
+    }
+
+
+    private boolean externalLinkIsDead(String url) {
+        LOG.fine("checking url: " + url + " ...");
+        if (linkResultCache.containsKey(url)) {
+            LOG.fine("response: HTTP " + linkResultCache.get(url) + " (CACHED) on " + url);
+            return linkResultCache.get(url) >= 400;
+        }
+
+        try {
+            final HttpURLConnection httpURLConnection = (HttpURLConnection) new URL(url).openConnection();
+            httpURLConnection.setRequestMethod("GET");
```

Comment:
since we only care about the status code, we could simply use `HEAD` instead of `GET`